text/0000-copy-closures.md

```diff
@@ -0,0 +1,170 @@
+`Clone`:
+
+```rust
+let mut x = 0;
+
+// `incr` mutates `x` and isn't a `move` closure,
+// so it's neither `Copy` nor `Clone`
+let incr = || { x += 1; };
+
+move_it(incr);
+move_it(incr); // ERROR: `print_incr` moved in the call above
+```
+
+`move` closures are only `Copy` or `Clone` if the values they capture are
+`Copy` or `Clone`:
+
+```rust
+let x = 5;
+
+// `x` is `Copy + Clone`, so `print_x` is `Copy + Clone`:
+let print_x = move || println!("{}", x);
+
+let foo = String::from("foo");
+// `foo` is `Clone` but not `Copy`, so `print_foo` is `Clone` but not `Copy`:
+let print_foo = move || println!("{}", foo);
+
+// Even closures which mutate variables are `Clone + Copy`
+// if their captures are `Clone + Copy`:
+let mut x = 0;
+
+// `x` is `Clone + Copy`, so `print_incr` is `Clone + Copy`:
+let print_incr = move || { println!("{}", x); x += 1; };
+move_it(print_incr);
+move_it(print_incr);
+move_it(print_incr);
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Closures are internally represented as structs which contain either values
+or references to the values of captured variables
+(`move` or non-`move` closures).
+A closure type implements `Clone` or `Copy` if and only if the all values in
+the closure's internal representation implement `Clone` or `Copy`:
+
+- Non-mutating non-`move` closures only contain immutable references
+(which are `Copy + Clone`), so these closures are `Copy + Clone`.
+
+- Mutating non-`move` closures contain mutable references, which are neither
+`Copy` nor `Clone`, so these closures are neither `Copy` nor `Clone`.
+
+- `move` closures contain values moved out of the enclosing scope, so these
+closures are `Clone` or `Copy` if and only if all of the values they capture
+are `Clone` or `Copy`.
+
+The internal implementation of `Clone` for non-`Copy` closures will resemble
+the basic implementation generated by `derive`, but the order in which values
+are `Clone`d will remain unspecified.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This feature increases the complexity of the language, as it will force users
+to reason about which variables are being captured in order to understand
+whether or not a closure is `Copy` or `Clone`.
+
+However, this can be mitigated through error messages which point to the
+specific captured variables that prevent a closure from satisfying `Copy` or
+`Clone` bounds.
+
+# Rationale and Alternatives
+[alternatives]: #alternatives
+
+It would be possible to implement `Clone` or `Copy` for a more minimal set of
+closures, such as only non-`move` closures, or non-mutating closures.
+This could make it easier to reason about exactly which closures implement
+`Copy` or `Clone`, but this would come at the cost of greatly decreased
+functionality.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- How can we provide high-quality, tailored error messages to indicate why a
```

Comment:
This is relatively easy. I don't think it has to be considered an unresolved question. However, I do think it would be valuable to give an example of the sort of message you would expect. =)